simulator/infra/replacement/cache_replacement.cpp

```diff
@@ -84,10 +85,123 @@ void LRUCacheInfo::erase_lru_element() {
     lru_hash.erase( lru_elem);
 }
 
+class Pseudo_LRUCacheInfo : public CacheReplacementInterface
+{
+    public:
+        explicit Pseudo_LRUCacheInfo( std::size_t ways);
+        void touch( std::size_t way) override;
+        std::size_t update() override;
+        std::size_t get_ways() const override { return ways; }
+
+        void set_to_erase( std::size_t ) override;
+        void allocate( std::size_t ) override;
+
+    private:
+        enum flags { Left = -2, Right = -1};
+
+        tree<int> lru_tree;
+        void construct_tree( tree<int>::iterator node_it, int depth);
+        void construct_leaf_layer();
+        int calculate_depth() const;
+        int side_of_the_sibling ( tree<int>::iterator node_it);
+        static int reverse_flag( int flag);
+
+        const std::size_t ways;
+        int leaf_iterator = 0;
+};
+
+Pseudo_LRUCacheInfo::Pseudo_LRUCacheInfo( std::size_t ways)
+    : ways( ways)
+{
+    lru_tree.insert( lru_tree.begin(), Left);
+    construct_tree( lru_tree.begin(), calculate_depth());
+}
+
+int Pseudo_LRUCacheInfo::calculate_depth() const
+{
+    int i = 0;
+        for ( int numerator = ways; numerator != 1; numerator = numerator/2)
+        {
+            if (numerator%2 != 0)
+                throw CacheReplacementException("Number of ways must be the power of 2!");
+            i++;
+        }
+    return i;
+}
+
+void Pseudo_LRUCacheInfo::construct_tree( tree<int>::iterator node_it, int depth)
+{
+    if (lru_tree.depth( node_it) < depth)
+    {
+        if ( lru_tree.depth( node_it) != depth - 1)
+        {
+            auto left_it = lru_tree.append_child( node_it, Left);
+            auto right_it = lru_tree.append_child( node_it, Left);
+            construct_tree( left_it, depth);
+            construct_tree( right_it, depth);
+        }
+        else
+        {
+            lru_tree.append_child(node_it, leaf_iterator++);
+            lru_tree.append_child(node_it, leaf_iterator++);
+        }
+    }
+}
+
+int Pseudo_LRUCacheInfo::reverse_flag( int flag)
+{
+    if ( flag == Left) {
+        return Right;
+        }
+    else
+        return Left;
+}
+
+int Pseudo_LRUCacheInfo::side_of_the_sibling ( tree<int>::iterator node_it)
+{
+    if  (lru_tree.index( node_it) == 0)
+        return Left;
+    return Right;
+}
+
+void Pseudo_LRUCacheInfo::touch( std::size_t way)
+{
+    if ( way >= ways)
```

Comment:
That's an antipattern of "defensive programming" — you do not trust the caller that it provides correct input.
It's better to drop check here, since we use unit tests to justify the inputs are correct.

Reply:
That one was not resolved...

Reply:
But I've dropped the check